File: 0680_valid_palindrome_II/valid_palindrome.go
package main

// Time: O(n)
// Space: O(1)
func validPalindrome(s string) bool {
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
		if s[i] != s[j] {
			i1, j1 := i, j-1
			i2, j2 := i+1, j
			for i1 < j1 && s[i1] == s[j1] {
				i1++
				j1--
			}
			for i2 < j2 && s[i2] == s[j2] {
				i2++
				j2--
			}
			return i1 >= j1 || i2 >= j2
		}
	}
	return true
}

func main() {

}

File: 0680_valid_palindrome_II/valid_palindrome.py
def valid_palindrome(s: str) -> bool:
    """
    Time: O(n)
    Space: O(n)
    """
    i, j = 0, len(s) - 1
    while i < j:
        if s[i] != s[j]:
            exclude_last = s[i: j]
            exclude_first = s[i + 1: j + 1]
            return is_palindrome(exclude_last) or is_palindrome(exclude_first)
        i += 1
        j -= 1
    return True


def is_palindrome(s: str):
    return s == s[::-1]
